File: qbopt/optimize/loopsimplify.py
from dataclasses import replace

from qbopt.model import ir
from qbopt.model import mir
from qbopt.analysis import ssa
from qbopt.analysis import loops
from qbopt.optimize import edges
from qbopt.model.passes import MIRTransform


class LoopSimplify(MIRTransform):
    name = "loopsimplify"

    def transform(self, body: mir.MirBody) -> mir.MirBody:
        return simplified(body)


def grouped(body: mir.MirBody, target: int, sources: frozenset[int]) -> mir.MirBody:
    destination = body.block(target)
    predecessors = loops.predecessors(body.blocks)
    if destination is None or not sources or not sources <= predecessors[target] or target == body.entry:
        return body
    parents = [block for block in body.blocks if block.at in sources]
    for parent in parents:
        if not parent.ops:
            return body
        last = parent.ops[-1]
        if last.kind is mir.Kind.BRANCH:
            if not edges.conditional(parent, target):
                return body
        elif last.kind is mir.Kind.JUMP:
            if parent.succ != (target,) or last.target != target:
                return body
        elif (
            parent.succ != (target,)
            or last.kind in {mir.Kind.CALL, mir.Kind.RETURN, mir.Kind.OPAQUE, mir.Kind.SWITCH, mir.Kind.ESCAPE}
            or last.barrier
        ):
            return body
    if any(set(phi.incoming) != predecessors[target] or phi.result.flags for phi in destination.phis):
        return body
    label = edges.fresh(body)
    serial = max((value.id for value in ssa.values(body)), default=-1) + 1
    versions = {}
    for value in ssa.values(body):
        versions[value.variable] = max(versions.get(value.variable, 0), value.version)
    bridge_phis = []
    target_phis = []
    for phi in destination.phis:
        incoming = {at: value for at, value in phi.incoming.items() if at in sources}
        values = set(incoming.values())
        if len(values) == 1:
            result = next(iter(values))
        else:
            variable = phi.result.variable
            versions[variable] = versions.get(variable, 0) + 1
            result = mir.Value(serial, label, variable=variable, version=versions[variable])
            serial += 1
            bridge_phis.append(mir.Phi(result, incoming))
        target_phis.append(
            replace(
                phi, incoming={**{at: value for at, value in phi.incoming.items() if at not in sources}, label: result}
            )
        )
    jump = mir.Op(
        label, ir.Operation.JUMP, "", (), (), kind=mir.Kind.JUMP, target=target, covers=(label, label), symbol=False
    )
    bridge = mir.MirBlock(label, tuple(bridge_phis), (jump,), (target,))
    changed = []
    for block in body.blocks:
        if block.at in sources:
            last = block.ops[-1]
            block = replace(
                block,
                succ=tuple(label if at == target else at for at in block.succ),
                ops=(*block.ops[:-1], replace(last, target=label) if last.target == target else last),
            )
        if block.at == target:
            block = replace(block, phis=tuple(target_phis))
        changed.append(block)
    return replace(body, blocks=(*changed, bridge))


def simplified(body: mir.MirBody) -> mir.MirBody:
    if loops.irreducible(body.blocks, body.entry):
        return body
    for original in loops.loops(body.blocks, body.entry):
        original = next(loop for loop in loops.loops(body.blocks, body.entry) if loop.header == original.header)
        candidate = body
        predecessors = loops.predecessors(candidate.blocks)
        outside = predecessors[original.header] - original.body
        if not outside:
            continue
        parent = candidate.block(next(iter(outside)))
        if parent is None:
            continue
        if len(outside) != 1 or parent.succ != (original.header,):
            candidate = grouped(candidate, original.header, outside)
            if candidate is body:
                continue
        if len(original.latches) != 1:
            changed = grouped(candidate, original.header, original.latches)
            if changed is candidate:
                continue
            candidate = changed
        current = next(
            loop for loop in loops.loops(candidate.blocks, candidate.entry) if loop.header == original.header
        )
        predecessors = loops.predecessors(candidate.blocks)
        exits = {
            at for block in candidate.blocks if block.at in current.body for at in block.succ if at not in current.body
        }
        for target in sorted(exits):
            sources = predecessors.get(target, frozenset()) & current.body
            if predecessors.get(target, frozenset()) - current.body:
                changed = grouped(candidate, target, sources)
                if changed is candidate:
                    break
                candidate = changed
        else:
            body = candidate
    return body

File: src/optimize/loopsimplify.rs
//! Port of `qbopt/optimize/loopsimplify.py`.
//!
//! Python's `is` on an unchanged body becomes `==`: a grouping always adds a
//! block, so a changed body never equals its input.

use std::rc::Rc;
use std::collections::{BTreeMap, BTreeSet};

use crate::analysis::{loops, ssa};
use crate::model::ir::Operation;
use crate::model::mir::{Kind, MirBlock, MirBody, Op, OpCode, OrderedMap, Phi, Value};
use crate::model::passes::MIRTransform;
use crate::optimize::edges;

pub(crate) struct LoopSimplify;

impl MIRTransform for LoopSimplify {
    fn class_name(&self) -> &'static str {
        "LoopSimplify"
    }

    fn name(&self) -> &str {
        "loopsimplify"
    }

    fn transform(&mut self, body: Rc<MirBody>) -> Result<Rc<MirBody>, String> {
        Ok(simplified(&body))
    }
}

pub(crate) fn grouped(body: &Rc<MirBody>, target: i64, sources: &BTreeSet<i64>) -> Rc<MirBody> {
    let destination = body.block(target);
    let predecessors = loops::predecessors(&body.blocks);
    let Some(destination) = destination else {
        return body.clone();
    };
    if sources.is_empty() || !sources.is_subset(&predecessors[&target]) || target == body.entry {
        return body.clone();
    }
    let parents = body.blocks.iter().filter(|block| sources.contains(&block.at));
    for parent in parents {
        let Some(last) = parent.ops.last() else {
            return body.clone();
        };
        if last.kind == Kind::Branch {
            if !edges::conditional(parent, target) {
                return body.clone();
            }
        } else if last.kind == Kind::Jump {
            if parent.succ != [target] || last.target != Some(target) {
                return body.clone();
            }
        } else if parent.succ != [target]
            || matches!(last.kind, Kind::Call | Kind::Return | Kind::Opaque | Kind::Switch | Kind::Escape)
            || last.barrier()
        {
            return body.clone();
        }
    }
    if destination.phis.iter().any(|phi| {
        phi.incoming.keys().copied().collect::<BTreeSet<_>>() != predecessors[&target] || phi.result.flags
    }) {
        return body.clone();
    }
    let label = edges::fresh(body);
    let mut serial = ssa::values(body).map(|value| i64::from(value.id)).max().unwrap_or(-1) + 1;
    let mut versions: BTreeMap<u32, u32> = BTreeMap::new();
    for value in ssa::values(body) {
        let known = versions.get(&value.variable).copied().unwrap_or(0);
        versions.insert(value.variable, known.max(value.version));
    }
    let mut bridge_phis = Vec::new();
    let mut target_phis = Vec::new();
    for phi in &destination.phis {
        let incoming = phi
            .incoming
            .iter()
            .filter(|(at, _)| sources.contains(at))
            .map(|(&at, &value)| (at, value))
            .collect::<OrderedMap<_, _>>();
        let values = incoming.values().copied().collect::<BTreeSet<_>>();
        let result = if values.len() == 1 {
            *values.iter().next().expect("one value")
        } else {
            let variable = phi.result.variable;
            let version = versions.get(&variable).copied().unwrap_or(0) + 1;
            versions.insert(variable, version);
            let result = Value { id: serial as u32, at: label, flags: false, variable, version };
            serial += 1;
            bridge_phis.push(Phi { result, incoming });
            result
        };
        let mut kept = phi
            .incoming
            .iter()
            .filter(|(at, _)| !sources.contains(at))
            .map(|(&at, &value)| (at, value))
            .collect::<OrderedMap<_, _>>();
        kept.insert(label, result);
        target_phis.push(Phi { incoming: kept, ..phi.clone() });
    }
    let mut jump = Op::new(label, OpCode::Operation(Operation::Jump), "", vec![], vec![]);
    jump.kind = Kind::Jump;
    jump.target = Some(target);
    jump.symbol = Some(false);
    let bridge = MirBlock::new(label, bridge_phis, vec![jump], vec![target]);
    let mut changed = Vec::new();
    for block in &body.blocks {
        let mut block = block.clone();
        if sources.contains(&block.at) {
            block.succ = block.succ.iter().map(|&at| if at == target { label } else { at }).collect();
            let last = block.ops.last_mut().expect("checked above");
            if last.target == Some(target) {
                last.target = Some(label);
            }
        }
        if block.at == target {
            block.phis = target_phis.clone();
        }
        changed.push(block);
    }
    changed.push(bridge);
    Rc::new(body.with_blocks(changed))
}

pub(crate) fn simplified(body: &Rc<MirBody>) -> Rc<MirBody> {
    let mut body = body.clone();
    if !loops::irreducible(&body.blocks, Some(body.entry)).is_empty() {
        return body;
    }
    for original in loops::loops(&body.blocks, Some(body.entry)) {
        let original = loops::loops(&body.blocks, Some(body.entry))
            .into_iter()
            .find(|loop_| loop_.header == original.header)
            .expect("StopIteration");
        let mut candidate = body.clone();
        let predecessors = loops::predecessors(&candidate.blocks);
        let outside = predecessors[&original.header].difference(&original.body).copied().collect::<BTreeSet<_>>();
        if outside.is_empty() {
            continue;
        }
        let Some(parent) = candidate.block(*outside.iter().next().expect("nonempty")) else {
            continue;
        };
        if outside.len() != 1 || parent.succ != [original.header] {
            candidate = grouped(&candidate, original.header, &outside);
            if Rc::ptr_eq(&candidate, &body) {
                continue;
            }
        }
        if original.latches.len() != 1 {
            let changed = grouped(&candidate, original.header, &original.latches);
            if Rc::ptr_eq(&changed, &candidate) {
                continue;
            }
            candidate = changed;
        }
        let current = loops::loops(&candidate.blocks, Some(candidate.entry))
            .into_iter()
            .find(|loop_| loop_.header == original.header)
            .expect("StopIteration");
        let predecessors = loops::predecessors(&candidate.blocks);
        let exits = candidate
            .blocks
            .iter()
            .filter(|block| current.body.contains(&block.at))
            .flat_map(|block| block.succ.iter().copied())
            .filter(|at| !current.body.contains(at))
            .collect::<BTreeSet<_>>();
        let empty = BTreeSet::new();
        let mut broke = false;
        for target in exits {
            let reaching = predecessors.get(&target).unwrap_or(&empty);
            let sources = reaching.intersection(&current.body).copied().collect::<BTreeSet<_>>();
            if !reaching.is_subset(&current.body) {
                let changed = grouped(&candidate, target, &sources);
                if Rc::ptr_eq(&changed, &candidate) {
                    broke = true;
                    break;
                }
                candidate = changed;
            }
        }
        if !broke {
            body = candidate;
        }
    }
    body
}

#[cfg(test)]
#[path = "loopsimplify_tests.rs"]
mod tests;
